### ISSUE TITLE
document issues with reentrancy in trio guest mode

### ISSUE THREAD

**Issue by njsmith** [python-trio/trio#1578]
Some host loops like to "reenter" the loop, e.g. in Qt stuff like `QDialog.exec`, a bunch of utility functions on `QFileDialog`, `QMenu.exec`, etc. See also:

- https://www.qt.io/blog/2010/02/23/unpredictable-exec
- https://stackoverflow.com/questions/35561182/why-should-nesting-of-qeventloops-be-avoided

Need to document what happens if you use these from inside a Trio function (not great – does it merely block Trio, or is it catastrophic? at first I thought catastrophic, but maybe it's merely the same as doing any blocking operation), and what to do instead (use alternate APIs, or push them out to run in the host event loop directly and wait for the result from Trio).

**Comment by njsmith** [comment]
@altendky reports running into this for real: https://gitter.im/python-trio/general?at=5edefa154c9b0f060d52834b

**Comment by altendky** [comment]
I actually avoided it in another spot before I hit it by using the dialog's `.show()` method instead of `.exec()` because we get to await signals unlike in Qt itself where using `.show()` would force a second function to connect the finished signal to.

**Comment by njsmith** [comment]
Yeah, I imagine an eventual `qtrio` library might also want to provide some convenience functions for common cases like showing dialogs / context menus + waiting for completion, and maybe also an `await run_sync_reentrant(some_sync_reentrant_function, *args)` helper that schedules the function on the Qt loop and then sends the outcome into Trio.

**Comment by richardsheridan** [comment]
I'm a bit late to this party, but this also shows up in `tkinter` whenever a modal dialog or widget needs to return a value. In my case this is the open/save file dialog. They "need" to do this because it allows the widget to look like a function call and return a value, a string with the chosen file's path inside.  The dialog is actually waiting on Tcl `vwait` which runs an event loop within the currently executing task until the required value is set, which is essentially trying to transform a subroutine into a coroutine. Such things should have been implemented in a continuation passing style, but alas.

I solved the issue by simply wrapping the reentrant call in a trio thread. This works thanks to `tkinter` having that specific cross-thread function calling solution, so the trio task sleeps properly and other trio tasks are correctly executed by the reentrant event loop. However, for this same reason "shove it in a thread" probably won't be a general solution for every host event loop, but at least it is something people could try until there is a general solution!
